File: android/app/src/main/java/com/nianxiang/app/particle/ParticleModels.kt
/** PUBLIC STUB — the depth→particle placement algorithm (ParticleDepth), sampling
 *  budget logic and particle data layout live in the private core module. The scene
 *  state types below are pure data shared with the UI layer and stay real. */
package com.nianxiang.app.particle

import org.json.JSONObject
import java.util.Base64

enum class ParticleAmbience { NONE, DUST, RAIN, SNOW }

sealed interface ParticleMode {
    data object Timeline : ParticleMode
    data object Loading : ParticleMode
    data object Chat : ParticleMode
    data class Condensing(val lines: List<String>) : ParticleMode
    data class Diary(val mood: String) : ParticleMode
    data object Hidden : ParticleMode
}

data class ParticleSceneState(
    val photoId: String? = null,
    val jpeg: ByteArray? = null,
    val depthJson: String? = null,
    val mode: ParticleMode = ParticleMode.Timeline,
)

data class ParticlePerformanceSample(
    val framesPerSecond: Float,
    val slowFrameRatio: Float,
    val sampledFrames: Int,
)

data class DepthPayload(
    val width: Int,
    val height: Int,
    val depth: ByteArray,
    val mask: ByteArray? = null,
    val background: ByteArray? = null,
    val backgroundDepth: ByteArray? = null,
) {
    val layered: Boolean
        get() = mask != null && background != null && backgroundDepth != null

    companion object {
        fun parse(json: String?): DepthPayload? {
            if (json.isNullOrBlank()) return null
            return runCatching {
                val value = JSONObject(json)
                val width = value.getInt("width")
                val height = value.getInt("height")
                require(width in 1..4096 && height in 1..4096)
                val pixels = Math.multiplyExact(width, height)
                val depth = decode(value.getString("depth"), pixels) ?: return@runCatching null
                if (!value.optBoolean("layered", false)) {
                    return@runCatching DepthPayload(width, height, depth)
                }
                val mask = decode(value.optString("mask"), pixels)
                val background = decode(value.optString("bg"), Math.multiplyExact(pixels, 3))
                val backgroundDepth = decode(value.optString("bgDepth"), pixels)
                if (mask == null || background == null || backgroundDepth == null) {
                    DepthPayload(width, height, depth)
                } else {
                    DepthPayload(width, height, depth, mask, background, backgroundDepth)
                }
            }.getOrNull()
        }

        private fun decode(value: String, expected: Int): ByteArray? {
            if (value.isBlank()) return null
            val decoded = runCatching { Base64.getDecoder().decode(value) }.getOrNull() ?: return null
            return decoded.takeIf { it.size == expected }
        }
    }
}

internal fun ambienceForMood(mood: String): ParticleAmbience = when {
    Regex("[雨思念忧伤愁怀]").containsMatchIn(mood) -> ParticleAmbience.RAIN
    Regex("[静雪冬凉安]").containsMatchIn(mood) -> ParticleAmbience.SNOW
    else -> ParticleAmbience.DUST
}

internal fun particleModeFor(
    sessionOpen: Boolean,
    userPresent: Boolean,
    hasPhoto: Boolean,
    phase: String,
    sessionTab: String,
    mood: String,
    lines: List<String>,
): ParticleMode = when {
    !userPresent || !sessionOpen -> ParticleMode.Timeline
    !hasPhoto -> ParticleMode.Loading
    phase == "condensing" -> ParticleMode.Condensing(lines.takeLast(8))
    (phase == "revealing" || phase == "done") && sessionTab == "diary" -> ParticleMode.Diary(mood)
    else -> ParticleMode.Chat
}

File: apple/Nianxiang/Particle/ParticleModels.swift
/// PUBLIC STUB — the depth→particle placement algorithm (ParticleDepth), sampling
/// budget logic and particle data layout live in the private core module. The scene
/// state types below are pure data shared with the UI layer and stay real.
import Foundation

enum ParticleAmbience {
    case none, dust, rain, snow
}

enum ParticleMode: Equatable {
    case timeline
    case loading
    case chat
    case condensing(lines: [String])
    case diary(mood: String)
    case hidden
}

struct ParticleSceneState {
    var photoId: String?
    var jpeg: Data?
    var depthJson: String?
    var mode: ParticleMode = .timeline
}

struct ParticlePerformanceSample {
    let framesPerSecond: Float
    let slowFrameRatio: Float
    let sampledFrames: Int
}

struct DepthPayload {
    let width: Int
    let height: Int
    let depth: [UInt8]
    let mask: [UInt8]?
    let background: [UInt8]?
    let backgroundDepth: [UInt8]?

    var layered: Bool { mask != nil && background != nil && backgroundDepth != nil }

    static func parse(_ json: String?) -> DepthPayload? {
        guard let json, !json.isEmpty,
              let data = json.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let width = object["width"] as? Int,
              let height = object["height"] as? Int,
              (1...4096).contains(width), (1...4096).contains(height) else { return nil }
        let pixels = width * height
        guard let depth = decode(object["depth"] as? String, expected: pixels) else { return nil }
        guard object["layered"] as? Bool == true else {
            return DepthPayload(width: width, height: height, depth: depth, mask: nil, background: nil, backgroundDepth: nil)
        }
        let mask = decode(object["mask"] as? String, expected: pixels)
        let background = decode(object["bg"] as? String, expected: pixels * 3)
        let backgroundDepth = decode(object["bgDepth"] as? String, expected: pixels)
        guard let mask, let background, let backgroundDepth else {
            return DepthPayload(width: width, height: height, depth: depth, mask: nil, background: nil, backgroundDepth: nil)
        }
        return DepthPayload(width: width, height: height, depth: depth, mask: mask, background: background, backgroundDepth: backgroundDepth)
    }

    private static func decode(_ value: String?, expected: Int) -> [UInt8]? {
        guard let value, !value.isEmpty, let decoded = Data(base64Encoded: value), decoded.count == expected else { return nil }
        return [UInt8](decoded)
    }
}

func ambienceForMood(_ mood: String) -> ParticleAmbience {
    if mood.range(of: "[雨思念忧伤愁怀]", options: .regularExpression) != nil { return .rain }
    if mood.range(of: "[静雪冬凉安]", options: .regularExpression) != nil { return .snow }
    return .dust
}

func particleModeFor(
    sessionOpen: Bool,
    userPresent: Bool,
    hasPhoto: Bool,
    phase: String,
    sessionTab: String,
    mood: String,
    lines: [String]
) -> ParticleMode {
    if !userPresent || !sessionOpen { return .timeline }
    if !hasPhoto { return .loading }
    if phase == "condensing" { return .condensing(lines: Array(lines.suffix(8))) }
    if (phase == "revealing" || phase == "done") && sessionTab == "diary" { return .diary(mood: mood) }
    return .chat
}

/// Deterministic xorshift RNG so tests can pin seeds; production paths use a random seed.
struct SeededRandom {
    private var state: UInt64

    init(seed: Int) {
        state = UInt64(bitPattern: Int64(seed)) &+ 0x9E3779B97F4A7C15
        if state == 0 { state = 0x1234_5678 }
    }

    init() {
        state = UInt64.random(in: 1...UInt64.max)
    }

    mutating func nextFloat() -> Float {
        state ^= state << 13
        state ^= state >> 7
        state ^= state << 17
        return Float(state >> 40) / Float(1 << 24)
    }

    mutating func nextInt(_ upperBound: Int) -> Int {
        Int(nextFloat() * Float(upperBound)).clamped(to: 0...(upperBound - 1))
    }
}

extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
